### PR TITLE
Feat/mutlithread

### PR DESCRIPTION
**Pull request by JeremyWesthead**
Adds multithreading for VCF parsing. While this is usually fast, there are pathological VCF cases with ~1000 base ref/alt pairs and ~250,000 rows which can make this take ~10mins
Note that such cases crashed my laptop when running with `gumpy`, so were never lightweight and likely would have taken considerable time previously.

Adds basic multithreading for such cases - reducing time taken to ~2mins in such cases. Overhead for usual cases is essentially unchanged (at maximum adding ~40ms) so saves time in an average case. Also has the option for changing the number of threads this runs on.
Only multithreads the VCF parsing though as this is the slowest part, other slow part is the genome difference, but this would require a considerable rework to make work

No idea how this performs on anything other than linux, it's possible this is horrible on OSX/Windows due to differing OS level thread support.

### CODE REVIEW

File: src/vcf.rs [hunk 1]
```diff
@@ -203,7 +149,7 @@ impl VCFFile {
                 }
             }
             // Add minor calls if the filter is passed or ignored, or specifcally just the MIN_FRS has failed
-            if ignore_filter || passed || filters.contains(&"MIN_FRS".to_string()) {
+            if ignore_filter || passed || record.filter.contains(&"MIN_FRS".to_string()) {
```
Comment (mcolpus) [inline src/vcf.rs]: Based on the comment string the last check should be something like `record.filter == vec![&"MIN_FRS".to_string()]`

**Review by mcolpus — APPROVED** [review]
Changes look good to me. I've not used this library before, but I assume it's the `par_iter()` which is using the multithreading step

File: src/vcf.rs (hunk 1 above)
Comment (JeremyWesthead) [reply src/vcf.rs]: Ah good catch. I think it actually needs to match the filters in the allowed filters list above

**Comment by JeremyWesthead** [comment]
Yeah the `par_iter()` gives an iterable which spreads out each element over a threadpool which is created in the background. It's a pretty neat way to push loops across threads without much boilerplate

File: src/difference.rs [hunk 2]
```diff
@@ -740,10 +750,10 @@ impl GeneDifference {
                                         for e in e.iter() {
                                             minor_evidence.push(e.clone());
                                         }
-                                        if cov.unwrap() > minor_cov {
+                                        if cov.is_some() && cov.unwrap() > minor_cov {
```
Comment (mcolpus) [inline src/difference.rs:753]: I think you could replace this with `cov > Some(minor_cov)` if you wanted based on how rust compares Options. Not that I think I'd recommend in this case

File: src/vcf.rs [hunk 3]
```diff
@@ -148,8 +148,29 @@ impl VCFFile {
...
+            let mut valid_filters = passed || ignore_filter;
+            if !valid_filters {
+                // Not passed filter and not ignored, so check if all filters are in allowed filters
+                let allowed_filters = [
+                    "MIN_FRS".to_string(),
+                    "MIN_DP".to_string(),
+                    "MIN_GCP".to_string(),
+                ];
+                // Auto-ignore if no filters are present (i.e filter = "." which is fail)
+                let mut all_allowed = !record.filter.is_empty();
```
Comment (mcolpus) [inline src/vcf.rs:162]: I'd replace with a statement like: `let all_allowed = record.filter.iter().all(|f| allowed_filters.contains(f));`

**Review by mcolpus — APPROVED** [review]

File: src/vcf.rs (hunk 3 above)
Comment (JeremyWesthead) [reply src/vcf.rs:162]: That doesn't cover the case of a 0 length vector unfortunately

Comment (mcolpus) [reply src/vcf.rs:162]: oh because you want to fail on an empty vector

**Pull request merged**